src/J2N/Collections/Generic/OrderedHashSet.cs

```diff
@@ -0,0 +1,2357 @@
+        /// different equality comparers, then uniqueness assumption between sets break.
+        /// </summary>
+        private static bool AreEqualityComparersEqual(OrderedHashSet<T> set1, OrderedHashSet<T> set2) =>
+            set1.EqualityComparer.Equals(set2.EqualityComparer);
+
+        /// <summary>
+        /// Checks if effective equality comparers are equal. This is used for algorithms that
+        /// require that both collections use identical hashing implementations for their entries.
+        /// </summary>
+        internal static bool EffectiveEqualityComparersAreEqual(OrderedHashSet<T> set1, OrderedHashSet<T> set2) =>
+            set1.EffectiveComparer.Equals(set2.EffectiveComparer);
+
+        #endregion Helper methods
+
+        #region Structural Equality
+
+        /// <summary>
+        /// Determines whether the specified object is structurally equal to the current set
+        /// using rules provided by the specified <paramref name="comparer"/>.
+        /// </summary>
+        /// <param name="other">The object to compare with the current object.</param>
+        /// <param name="comparer">The <see cref="IEqualityComparer"/> implementation to use to determine
+        /// whether the current object and <paramref name="other"/> are structurally equal.</param>
+        /// <returns><c>true</c> if <paramref name="other"/> is structurally equal to the current set;
+        /// otherwise, <c>false</c>.</returns>
+        /// <exception cref="ArgumentNullException">If <paramref name="comparer"/> is <c>null</c>.</exception>
+        public virtual bool Equals(object? other, IEqualityComparer comparer)
+            => SetEqualityComparer<T>.Equals(this, other, comparer);
+
+        /// <summary>
+        /// Gets the hash code representing the current set using rules specified by the
+        /// provided <paramref name="comparer"/>.
+        /// </summary>
+        /// <param name="comparer">The <see cref="IEqualityComparer"/> implementation to use to generate
+        /// the hash code.</param>
+        /// <returns>A hash code representing the current set.</returns>
+        public virtual int GetHashCode(IEqualityComparer comparer)
+            => SetEqualityComparer<T>.GetHashCode(this, comparer);
+
+        /// <summary>
+        /// Determines whether the specified object is structurally equal to the current set
+        /// using rules similar to those in the JDK's AbstractSet class. Two sets are considered
+        /// equal when they both contain the same objects (in any order).
+        /// </summary>
+        /// <param name="obj">The object to compare with the current object.</param>
+        /// <returns><c>true</c> if the specified object implements <see cref="ISet{T}"/>
+        /// and it contains the same elements; otherwise, <c>false</c>.</returns>
+        /// <seealso cref="Equals(object, IEqualityComparer)"/>
+        public override bool Equals(object? obj)
```

Comment:
I think we can get a performance boost (not just for this collection, but for all of them) by adding a special case when `obj` is the current type and taking full advantage of the `Enumerator` by casting to `HashSet<T>`. We may even be able to do better by basing the equality on the backing arrays rather than this collection's enumerator.

Here is the implementation from `AbstractSet<E>`:

https://github.com/openjdk/jdk8/blob/master/jdk/src/share/classes/java/util/AbstractSet.java#L68-L101

Note that this isn't overridden in `LinkedHashSet<E>`.